pkg/tls/self_signed_cert.go

```diff
@@ -0,0 +1,205 @@
+
+// Package gencert generates self-signed TLS certificates.
+package tls
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"log/slog"
+	"math/big"
+	"os"
+	"time"
+)
+
+const (
+	caOrganization      = "F5 Inc. CA"
+	certOrganization    = "F5 Inc."
+	certFilePermissions = 0o600
+	keyFilePermissions  = 0o600
+)
+
+type certReq struct {
+	template   *x509.Certificate
+	parent     *x509.Certificate
+	publicKey  *ecdsa.PublicKey
+	privateKey *ecdsa.PrivateKey
+}
+
+func genCert(req *certReq) (*x509.Certificate, []byte) {
+	certBytes, createCertErr := x509.CreateCertificate(
+		rand.Reader,
+		req.template,
+		req.parent,
+		req.publicKey,
+		req.privateKey,
+	)
+
+	if createCertErr != nil {
+		slog.Error("Failed to generate certificate", "error", createCertErr)
+		return &x509.Certificate{}, []byte{}
+	}
+
+	cert, parseCertErr := x509.ParseCertificate(certBytes)
+	if parseCertErr != nil {
+		slog.Error("Failed to parse certificate")
+		return &x509.Certificate{}, []byte{}
+	}
+
+	b := pem.Block{Type: "CERTIFICATE", Bytes: certBytes}
+	certPEM := pem.EncodeToMemory(&b)
+
+	return cert, certPEM
+}
+
+func GenerateCA(now time.Time, caCertPath string) (*x509.Certificate, *ecdsa.PrivateKey, error) {
+	// Generate key pair for the CA
+	caKeyPair, caKeyErr := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if caKeyErr != nil {
+		return &x509.Certificate{}, &ecdsa.PrivateKey{}, fmt.Errorf("failed to generate CA private key: %w", caKeyErr)
+	}
+
+	// Create CA certificate template
+	caTemplate := x509.Certificate{
+		SerialNumber:          big.NewInt(1),
+		Subject:               pkix.Name{Organization: []string{certOrganization}},
+		NotBefore:             now.Add(-time.Minute),
+		NotAfter:              now.AddDate(1, 0, 0), // 1 year
+		KeyUsage:              x509.KeyUsageCertSign | x509.KeyUsageCRLSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+		MaxPathLen:            1,
+	}
+
+	// CA is self signed
+	caRequest := certReq{
+		template:   &caTemplate,
+		parent:     &caTemplate,
+		publicKey:  &caKeyPair.PublicKey,
+		privateKey: caKeyPair,
+	}
+
+	caCert, caCertPEM := genCert(&caRequest)
+	if len(caCertPEM) == 0 {
+		slog.Error("Error generating certificate authority")
+	}
+
+	// Write the CA certificate to a file
+	slog.Debug("About to write CA file", "path", caCertPath)
+	writeCAErr := os.WriteFile(caCertPath, caCertPEM, certFilePermissions)
+	if writeCAErr != nil {
+		return &x509.Certificate{}, &ecdsa.PrivateKey{}, fmt.Errorf(
+			"failed to write ca file: %w",
+			writeCAErr,
+		)
+	}
+
+	return caCert, caKeyPair, nil
+}
+
+// nolint: revive
+func GenerateServerCert(hostnames []string, caPath, certPath, keyPath string) (existingCert bool, err error) {
+	// Check for and return existing cert if it already exists
```

Comment:
If I was to use this function I would assume that it would create certs. If certs already exist mayeb if should return a file already exists error?